### PR TITLE
src: build: Adding verbose mode.

### PR DESCRIPTION
**Pull request by aquilamacedo**
This commit adds support for the verbose parameter within `kw build`. The verbose parameter gives details of the commands that are executed behind the scenes.

Note: This is part of the issue: #179

### CODE REVIEW

**Comment by aquilamacedo** [comment]
Hi @rodrigosiqueira. 

I've started implementing the verbose parameter in `kw build`, do you have any recommendations for something I can change/add?

File: src/build.sh [hunk 1]
```diff
@@ -102,7 +106,7 @@ function kernel_build()
   fi
 
   if [[ -n "$clean" ]]; then
-    build_clean "$flag"
+    build_clean "$flag" "$output_kbuild_flag"
```
Comment (rodrigosiqueira) [inline src/build.sh:109]: This change and the modification to `build_clean` should be in its own commit.

File: src/build.sh [hunk 2]
```diff
@@ -118,8 +122,7 @@ function kernel_build()
   fi
 
   if [[ -n "$menu_config" ]]; then
-    command="make -j ${llvm}ARCH=${platform_ops} ${menu_config}${output_kbuild_flag}"
-    cmd_manager "$flag" "$command"
+    build_menu_config "$flag" "$output_kbuild_flag" "$menu_config" "$platform_ops" "$llvm"
```
Comment (rodrigosiqueira) [inline src/build.sh:125]: This modification should have a dedicated commit to it.

File: src/build.sh [hunk 3]
```diff
@@ -142,11 +145,12 @@ function kernel_build()
   fi
 
   if [[ -n "$doc_type" ]]; then
-    command="make ${optimizations} ${doc_type}${output_path}${output_kbuild_flag}"
-    cmd_manager "$flag" "$command"
-    return
+    build_doc "$flag" "$output_kbuild_flag" "$optimizations" "$doc_type" "$output_path"
```
Comment (rodrigosiqueira) [inline src/build.sh:148]: This change should have a dedicated commit.

File: src/build.sh [hunk 4]
```diff
@@ -168,6 +172,41 @@ function kernel_build()
...
+# This function builds kernel-doc, by default it will create htmldocs.
+function build_doc()
+{
+  local flag="$1"
+  local env_path="$2"
+  local optimizations="$3"
+  local doc_type="$4"
+  local output_path="$5"
+  local cmd
+  say "$doc_type"
```
Comment (rodrigosiqueira) [inline src/build.sh:203]: Add space between the variable declaration and `say`.

File: src/build.sh [hunk 5]
```diff
@@ -352,6 +402,10 @@ function parse_build_options()
       --)
         shift
         ;;
+      TEST_MODE)
+        options_values['TEST_MODE']='TEST_MODE'
+        shift
+        ;;
```
Comment (rodrigosiqueira) [inline src/build.sh:408]: I don't think we need this.

File: src/build.sh [hunk 6]
```diff
@@ -376,6 +430,7 @@ function build_help()
     '  build (-S | --cpu-scaling) <percentage> - Scale CPU usage by factor' \
     '  build (--ccache) - Enable use of ccache' \
     '  build (-w | --warnings) [warning_levels] - Enable warnings' \
+    '  build (-v | --verbose) - Show a detailed output' \
```
Comment (rodrigosiqueira) [inline src/build.sh:433]: You also need to update the man doc.

**Review by rodrigosiqueira — CHANGES_REQUESTED** [review]

**Comment by aquilamacedo** [comment]
Hi @rodrigosiqueira. 

I dedicated the commits to another pull request. When you merge, I'll update that pull request to continue the verbose option.

File: src/build.sh (hunk 5 above)
Comment (aquilamacedo) [reply src/build.sh]: We really don't need that, I ended up making a mistake.

**Comment by aquilamacedo** [comment]
Hi @rodrigosiqueira. :)

I started to add verbose mode in some parts of the build, but I'm still implementing it for other parameters.

**Comment by aquilamacedo** [comment]
Hi @rodrigosiqueira.

The `--verbose` option works for all **build** options except in the `--info` parameter. I wondered if it is worth implementing, as the only commands running in the build_info function are the commands created by **get_kernel_release()** and **get_kernel_versions()**. What do you think?

File: documentation/man/features/build.rst [hunk 7]
```diff
@@ -16,6 +16,8 @@ SYNOPSIS
 | *kw* (*b* | *build*) [\--llvm] [\--alert=(s | v | (sv | vs) | n)]
 | *kw* (*b* | *build*) [(-c | \--clean)] [\--alert=(s | v | (sv | vs) | n)]
 | *kw* (*b* | *build*) [(-f | \--full-cleanup)] [\--alert=(s | v | (sv | vs) | n)]
+| *kw* (*b* | *build*) [(-v | \--verbose)] [\--alert=(s | v | (sv | vs) | n)]
```
Comment (rodrigosiqueira) [inline documentation/man/features/build.rst:19]: Add `--verbose` in the `bash_autocomplete` and `_kw`.

File: documentation/man/features/env.rst [hunk 8]
```diff
@@ -11,6 +11,7 @@ SYNOPSIS
 | *kw* (*env*) [(-d | \--destroy <NAME>)]
 | *kw* (*env*) [(-l | \--list)]
 | *kw* (*env*) [(-e | \--exit-env)]
+| *kw* (*env*) [(-v | \--verbose)]
```
Comment (rodrigosiqueira) [inline documentation/man/features/env.rst:14]: Add `--verbose` in the bash_autocomplete and `_kw`.

**Review by rodrigosiqueira — CHANGES_REQUESTED** [review]

**Comment by rodrigosiqueira** [comment]
> Hi @rodrigosiqueira.
> 
> The `--verbose` option works for all **build** options except in the `--info` parameter. I wondered if it is worth implementing, as the only commands running in the build_info function are the commands created by **get_kernel_release()** and **get_kernel_versions()**. What do you think?

I think it is worth it. Check get_kernel_* function, and you will see that we are forcing SILENT.

**Comment by rodrigosiqueira** [comment]
Hi @aquilamacedo , could you finish this PR?

**Comment by aquilamacedo** [comment]
> Hi @aquilamacedo , could you finish this PR?

yes, I'm almost done updating _kw.

File: documentation/man/features/build.rst (hunk 7 above)
Comment (magalilemes) [reply documentation/man/features/build.rst:19]: It seems like `--alert` is a legacy option, which no longer works now. So no need to add it here.

File: src/build.sh [hunk 9]
```diff
@@ -31,6 +31,7 @@ function build_kernel_main()
   local llvm
   local env_name
   local clean
+  local flag
```
Comment (magalilemes) [inline src/build.sh:34]: This variable has already been declared.

File: src/kw_env.sh
Comment (magalilemes) [inline src/kw_env.sh:1]: Just so that I can understand this better: is `flag`, in this case, a global variable? If so, perhaps declare it that way.

Comment (aquilamacedo) [reply src/kw_env.sh:1]: Yes, the variable is being used globally as it is used in several functions. do you think it would be necessary to declare it as global?

File: src/build.sh (hunk 9 above)
Comment (aquilamacedo) [reply src/build.sh:34]: ok, i'm fixing this.

File: src/build.sh [hunk 10]
```diff
@@ -119,6 +121,8 @@ function build_kernel_main()
     return "$?"
   fi
 
+  flag=${flag:-'SILENT'}
```
Comment (rodrigosiqueira) [inline src/build.sh:124]: This should be at the beginning of this function.

File: src/build.sh [hunk 11]
```diff
@@ -353,6 +366,10 @@ function parse_build_options()
         options_values['FULL_CLEANUP']=1
         shift
         ;;
+      --verbose | -v)
+        options_values['VERBOSE']=1
+        shift
```
Comment (rodrigosiqueira) [inline src/build.sh:371]: Add a simple unit test for this new parameter (same idea for all the commits in this PR).

File: documentation/man/features/build.rst [hunk 12]
```diff
@@ -83,6 +84,9 @@ OPTIONS
   build process, including configuration files and script output. This option also
   considers whether or not the user is using kw env.
 
+-v, \--verbose:
+  Verbose mode allows the user to see the commands executed under the hood.
```
Comment (rodrigosiqueira) [inline documentation/man/features/build.rst:88]: Let's use something more formal in the documentation. Something like this:

Verbose mode. Causes kw to print debugging messages about its progress. This is helpful in debugging issues.

(Same idea for all commits)

File: src/kw_env.sh
Comment (rodrigosiqueira) [reply src/kw_env.sh:1]: flag should not be global. It should be local.

File: src/kw_env.sh [hunk 13]
```diff
@@ -76,14 +81,17 @@ function use_target_env()
       # Check if the config file exists before trying to remove it.
       [[ ! -f "${local_kw_configs}/${config}.config" ]] && continue
 
-      mv "${local_kw_configs}/${config}.config" "$tmp_trash"
+      cmd="mv ${local_kw_configs}/${config}.config $tmp_trash"
```
Comment (rodrigosiqueira) [inline src/kw_env.sh:84]: since you are making this change, please add `{}` around `tmp_trash`.

File: src/backup.sh [hunk 14]
```diff
@@ -14,19 +14,23 @@ decompress_path="$KW_CACHE_DIR/tmp-kw-backup"
 # recieves an argument and call other functions based on that.
 function backup()
 {
+  local flag
+
```
Comment (rodrigosiqueira) [inline src/backup.sh:18]: Add `flag=${flag:-'SILENT'}`.

File: documentation/man/features/mail.rst [hunk 15]
```diff
@@ -12,6 +12,7 @@ SYNOPSIS
 | *kw mail* (-l | \--list)
 | *kw mail* \--verify [\--local | \--global]
 | *kw mail* \--template[=<template>] [-n | \--no-interactive] [\--local | \--global] [-f | \--force] [(<config> <value>)...]
+| *kw mail* (-V | \--verbose)
```
Comment (rodrigosiqueira) [inline documentation/man/features/mail.rst:15]: In this case, let's drop the -V option

File: src/mail.sh [hunk 16]
```diff
@@ -23,6 +23,8 @@ declare -gr email_regex='[A-Za-z0-9_\.-]+@[A-Za-z0-9_-]+(\.[A-Za-z0-9]+)+'
 #shellcheck disable=SC2119
 function mail_main()
 {
+  local flag
+
```
Comment (rodrigosiqueira) [inline src/mail.sh:27]: Add `flag=${flag:-'SILENT'}` or similar

**Review by rodrigosiqueira — CHANGES_REQUESTED** [review]

File: documentation/man/features/mail.rst (hunk 15 above)
Comment (aquilamacedo) [reply documentation/man/features/mail.rst:15]: What do you mean when you say 'drop the -V option'? Are you suggesting removing the short option (-V) for verbose, leaving only the long option (--verbose)?

Comment (rodrigosiqueira) [reply documentation/man/features/mail.rst:15]: just remove the short option `-V` for the mail feature.

**Comment by rodrigosiqueira** [comment]
Merged! Thanks
